evm_arithmetization/src/prover.rs

```diff
@@ -249,6 +340,1098 @@ type ProofWithMemCaps<F, C, H, const D: usize> = (
+    // Quotient polynomials. They are already chunked in `degree` pieces.
+    let alphas = challenger.get_n_challenges(config.num_challenges);
+    let quotient_polys = all_quotient_polys::<F, P, C, D>(
+        all_stark,
+        &trace_poly_values_sorted,
+        &trace_commitment,
+        &auxiliary_columns_sorted,
+        &auxiliary_commitment,
+        &auxiliary_columns,
+        None,
+        &ctl_data_per_table,
+        alphas.clone(),
+        config,
+    );
+
+    // We compute the Field Merkle Tree of all quotient polynomials.
+    let quotient_polys_sorted: Vec<_> = Table::all_sorted()
+        .iter()
+        .map(|&table| quotient_polys[*table].clone())
+        .collect();
+    let quotient_polys_sorted_flat: Vec<_> = quotient_polys_sorted
+        .clone()
+        .into_iter()
+        .flatten()
+        .collect();
+    let num_quotient_polys = quotient_polys_sorted_flat.len();
+    let quotient_commitment = timed!(
+        timing,
+        "compute quotient commitments",
+        BatchFriOracle::<F, C, D>::from_coeffs(
+            quotient_polys_sorted_flat,
+            rate_bits,
+            false,
+            cap_height,
+            timing,
+            &vec![None; num_quotient_polys],
+        )
+    );
+    challenger.observe_cap(&quotient_commitment.batch_merkle_tree.cap);
+
+    let zeta = challenger.get_extension_challenge::<D>();
+
+    // To avoid leaking witness data, we want to ensure that our opening locations,
+    // `zeta` and `g * zeta`, are not in our subgroup `H`. It suffices to check
+    // `zeta` only, since `(g * zeta)^n = zeta^n`, where `n` is the order of
+    // `g`.
+    let degree_bits = trace_commitment.degree_bits[0];
+    let g = F::primitive_root_of_unity(degree_bits);
+    ensure!(
+        zeta.exp_power_of_2(degree_bits) != F::Extension::ONE,
+        "Opening point is in the subgroup."
+    );
+
+    let all_fri_instances = all_fri_instance_info::<F, C, D>(
+        all_stark,
+        zeta,
+        &trace_poly_values_sorted,
+        &auxiliary_columns_sorted,
+        &quotient_polys_sorted,
+        &ctl_data_per_table,
+        config,
+    );
+
+    // Get the FRI openings and observe them.
+    // Compute all openings: evaluate all committed polynomials at `zeta` and, when
+    // necessary, at `g * zeta`.
+    let openings = all_openings(
+        all_stark,
+        &trace_poly_values_sorted,
+        &trace_commitment,
+        &auxiliary_columns_sorted,
+        &auxiliary_commitment,
+        &quotient_polys_sorted,
+        &quotient_commitment,
+        &ctl_data_per_table,
+        zeta,
+        config,
+    );
+
+    for opening in openings.iter() {
+        challenger.observe_openings(&opening.to_fri_openings());
+    }
+
+    let initial_merkle_trees = [
+        &trace_commitment,
+        &auxiliary_commitment,
+        &quotient_commitment,
+    ];
+
+    let mut degree_bits_squashed = Table::all_degree_logs().to_vec();
+    degree_bits_squashed.dedup();
```

Comment:
why not a `HashSet`?

Reply:
We need a slice as argument for `prove_openings` so I decided to go with a vec directly.